src/cswinrt/strings/WinRT.cs

```diff
@@ -500,34 +532,100 @@ public unsafe ObjectReference<I> _ActivateInstance<I>()
     }
 
-    internal sealed class ActivationFactory<T> : BaseActivationFactory
+    internal static class ActivationFactory
     {
-        public ActivationFactory() : base(typeof(T).Namespace, typeof(T).FullName) { }
+        public static FactoryObjectReference<IActivationFactoryVftbl> Get(string typeName)
+        {
+            // Prefer the RoGetActivationFactory HRESULT failure over the LoadLibrary/etc. failure
+            int hr;
+            FactoryObjectReference<IActivationFactoryVftbl> factory;
+            (factory, hr) = WinrtModule.GetActivationFactory<IActivationFactoryVftbl>(typeName, InterfaceIIDs.IActivationFactory_IID);
+            if (factory != null)
+            {
+                return factory;
+            }
 
-        static readonly ActivationFactory<T> _factory = new ActivationFactory<T>();
-        public static new I AsInterface<I>() => _factory.Value.AsInterface<I>();
-        public static ObjectReference<I> As<I>() => _factory._As<I>();
-        public static IObjectReference As(Guid iid) => _factory._As(iid);
+            var moduleName = typeName;
+            while (true)
+            {
+                var lastSegment = moduleName.LastIndexOf(".", StringComparison.Ordinal);
+                if (lastSegment <= 0)
+                {
+                    Marshal.ThrowExceptionForHR(hr);
+                }
+                moduleName = moduleName.Remove(lastSegment);
 
+                DllModule module = null;
+                if (DllModule.TryLoad(moduleName + ".dll", out module))
+                {
+                    (factory, hr) = module.GetActivationFactory(typeName);
+                    if (factory != null)
+                    {
+                        return factory;
+                    }
+                }
+            }
+        }
+
 #if NET
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2091:RequiresUnreferencedCode", 
-            Justification = "No members of the generic type are dynamically accessed in this code path.")]
-#endif
-        public static ObjectReference<I> ActivateInstance<
+        public static FactoryObjectReference<I> Get<
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicConstructors | DynamicallyAccessedMemberTypes.PublicFields)]
+#else
+        public static ObjectReference<I> Get<
+#endif
+        I>(string typeName, Guid iid)
+        {
+            // Prefer the RoGetActivationFactory HRESULT failure over the LoadLibrary/etc. failure
+            int hr;
+            FactoryObjectReference<I> factory;
+            (factory, hr) = WinrtModule.GetActivationFactory<I>(typeName, iid);
+            if (factory != null)
+            {
 #if NET
-            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.None)]
-#endif
-            I>() => _factory._ActivateInstance<I>();
+                return factory;
+#else
+                using (factory)
+                {
+                    return factory.As<I>(iid);

```

Comment:
Is this else clause actually needed?  Given you are passing IID to GetActivationFactory and using the same I, does this do something more?